af/fastpath/fastpath/core.py

```diff
@@ -0,0 +1,574 @@
+    )
+    summary["scores"] = scores
+    return summary
+
+
+@metrics.timer("trivial_id")
+def trivial_id(msm):
+    """Generate a trivial id of the measurement to allow upsert if needed
+    - 32-bytes hexdigest
+    - Deterministic / stateless with no DB interaction
+    - Malicious/bugged msmts with collisions on report_id/input/test_name lead
+    to different hash values avoiding the collision
+    - Malicious/duplicated msmts that are semantically identical to the "real"
+    one lead to harmless collisions
+    """
+    # NOTE: we want the id to stay the same when a msmt is fetched from SSH
+    # and from a can on ooni-data-private/canned
+    # Implementing a rolling hash without ujson.dumps is 2x faster
+    # A rolling hash on only the first 2 levels of the dict is 10x faster
+    #
+    # Same output with Python's json
+    VER = "00"
+    msm_jstr = ujson.dumps(msm, sort_keys=True, ensure_ascii=False).encode()
+    tid = VER + hashlib.shake_128(msm_jstr).hexdigest(15)
+    return msm_jstr, tid
+
+
+def generate_filename(tid):
+    """Generate filesystem-safe filename from a measurement
+    """
+    return tid + ".json.lz4"
+
+
+@metrics.timer("writeout_measurement")
+def writeout_measurement(msm_jstr, fn, update):
+    """Safely write measurement to disk
+    """
+    # Different processes might be trying to write the same file at the same
+    # time due to naming collisions. Use a safe tmpfile and atomic link
+
+    suffix = ".{}.tmp".format(os.getpid())
+    with NamedTemporaryFile(suffix=suffix, dir=conf.msmtdir) as f:
+        with lz4frame.open(f, "w") as lzf:
+            lzf.write(msm_jstr)
+            # os.fsync(lzf.fileno())
+
+            final_fname = conf.msmtdir.joinpath(fn)
+            try:
+                os.link(f.name, final_fname)
+            except FileExistsError:
+                if update:
+                    # update access time - used for cache cleanup
+                    # no need to overwrite the file
+                    os.utime(final_fname)
+                else:
+                    log.info("Refusing to overwrite %s", final_fname)
+                    metrics.incr("report_id_input_file_collision")
+                    os.utime(final_fname)
+
+    metrics.incr("wrote_uncompressed_bytes", len(msm_jstr))
+
+
+def msm_processor(queue):
+    """Measurement processor worker
+    """
+    db.setup()
+    while True:
+        msm_tup = queue.get()
+
+        if msm_tup is None:
+            return
+
+        with metrics.timer("full_run"):
+            try:
+                msm_jstr, measurement = msm_tup
+                if measurement is None:
+                    measurement = ujson.loads(msm_jstr)
+                msm_jstr, tid = trivial_id(measurement)
+                fn = generate_filename(tid)
+                writeout_measurement(msm_jstr, fn, conf.update)
+                matches = match_fingerprints(measurement)
+                summary = score_measurement(measurement, matches)
```

Comment:
These two functions are applied to every measurements no matter the type. I think you should consider having some logic in place where the processing logic applied to measurements is dependent on the `test_name` key value. For example you probably don't want to be doing the `match_fingerprints` for middlebox test results and the `score_measurement` logic for `web_connectivity` and `facebook_messenger` will be different.

Reply:
The idea is to apply them to all msmts for consistency. Then, inside each function we'll implement more heuristics over time to handle each msmt type. Worst case we simply return early.